Persist and show openscap data

moolitayer opened the pull request:
Adds OpenSCAP entities and persists them during container smart state analysis. 

Container Images page addition:
![c_image_with_scap](https://cloud.githubusercontent.com/assets/3010449/13772475/243516b2-ea9d-11e5-907a-89d68478ea15.png)

This is how it looks if there are no openscap results:
![container_image_no_openscap_result](https://cloud.githubusercontent.com/assets/3010449/13772280/f603d6f8-ea9b-11e5-8c6a-e32d28ba1fc0.png)

List of OpenSCAP rule results:
![openscap_rule_results](https://cloud.githubusercontent.com/assets/3010449/13555055/bf924c36-e3bf-11e5-8fdd-750dafadec3e.png)

Singular OpenSCAP rule result:
![openscap_rule_result](https://cloud.githubusercontent.com/assets/3010449/13555068/e190547c-e3bf-11e5-8a80-fa93351e51c3.png)

Downloading the OpenSCAP HTML:
![Uploading Screenshot from 2016-03-15 11-05-33.png…]()

OpenSCAP HTML example:
![openscap_html](https://cloud.githubusercontent.com/assets/3010449/13555118/94ec9bd4-e3c0-11e5-97e6-2fe45c297c75.png)

Part two:
![openscap_html2](https://cloud.githubusercontent.com/assets/3010449/13555121/ab5c6d72-e3c0-11e5-83df-9b682bc732db.png)

moolitayer commented:
@miq-bot add_label providers/container, wip, smart_state
fyi @simon3z

isimluk commented:
@moolitayer I have been thinking about this. At this first step focus seems to be only on vulnerability assessment (that is also my takeaway from `ContainerImage has_one openscap_result`). Thus, if agree, that the first stage is to focus only on vulnerabilities, then perhaps it makes sense to not persist items of the scan that has passed. Usually, you would care only about failures when doing the vulnerability assessment. 

What do you think?

moolitayer commented:
> @moolitayer I have been thinking about this. At this first step focus seems to be only on vulnerability assessment (that is also my takeaway from ContainerImage has_one openscap_result). Thus, if agree, that the first stage is to focus only on vulnerabilities, then perhaps it makes sense to not persist items of the scan that has passed. Usually, you would care only about failures when doing the vulnerability assessment.
> 
> What do you think?

@isimluk ManageIQ allows administrators to define custom policies based on object properties. 
It seems logical for an administrator to mark an image as non compliant unless there is a record with it's openscap id and result = 'pass' . that way he can be sure that the check actually happened. makes sense?

isimluk commented:
Ah, I see. Thanks @moolitayer!

moolitayer commented:
This pr now only handles persisting and showing openscap data.
@simon3z @jrafanie @h-kataria please review

moolitayer commented:
@miq-bot remove_label wip
@miq-bot add_label ui

moolitayer commented:
cc @roliveri

jrafanie commented on app/models/openscap_result.rb:
```diff
@@ -0,0 +1,68 @@
+  has_one    :binary_blob,           :dependent => :destroy, :autosave => true, :as => :resource
+  has_many   :openscap_rule_results, :dependent => :destroy, :autosave => true
+
+  OpenSCAP.oscap_init
```
What does this do?  It's problematic to do expensive operations when loading a class, where expensive could be platform specific code, loading many/slow/high memory libraries, connecting to a database/remote service,  or anything similar.

moolitayer replied:
Looks like [this](https://github.com/OpenSCAP/openscap/blob/a50bbd9481dbc4aa1f3fa83902e665652f39d31c/src/common/oscapxml.c#L103) through [this](https://github.com/OpenSCAP/openscap/blob/a50bbd9481dbc4aa1f3fa83902e665652f39d31c/src/common/oscapxml.c#L103) wrapper. maybe @isimluk can elaborate.  

Anyway @jrafanie is there a way to do this we can live with? (Lazy initialization/platform hook)

I will also make sure that this is actually needed

jrafanie replied:
@moolitayer if this is required at container analysis time, the analysis can do it at the very beginning.  Requiring this class shouldn't have unexpected side effects on the system.  If someone isn't analyzing containers, is the overhead of running this code in terms of memory usage, platform compatibility (does it work on linux/mac), wall time adversely affected by this change?

isimluk replied:
`OpenSCAP.oscap_init` has very small foot print. However, I would also vote for running it only just before specific openscap operation (like creating ARF). The other operations, actually, may not need this init to be called. If you can, please run `OpenSCAP.oscap_cleanup()` afterwards.

chessbyte replied:
I would suggest we do something similar to `FileDepotFtp#with_connection` method.  Maybe something like the following:
```ruby
def with_open_scap
  raise "no block given" unless block_given?
     begin
      OpenSCAP.oscap_init
      yield
    ensure
      OpenSCAP.oscap_cleanup
    end
end
```

moolitayer replied:
done thanks!

jrafanie commented on db/migrate/20160220123313_add_open_scap_result_entities.rb:
```diff
@@ -0,0 +1,25 @@
+    end
+
+    create_table :openscap_rule_results do |t|
+      t.belongs_to :openscap_result
```
I don't remember if we have to specify type => bigint on association methods like belongs_to.  Either way, we specify it above and not here so let's make it consistent.

moolitayer replied:
Fixed thanks!

jrafanie commented on app/models/openscap_result.rb:
```diff
@@ -0,0 +1,68 @@
+    end
+  end
+
+  def parse_arf(arf_path)
```
Silly question, what's an arf?

isimluk replied:
Asset Reporting Format. Also known as Result DataStream. Standardized by NIST http://scap.nist.gov/specifications/arf/ It is the single format, you can fit all the information about the scan in.

moolitayer replied:
Added a comment to that.

jrafanie commented on app/models/openscap_result.rb:
```diff
@@ -0,0 +1,68 @@
+    rs = arf.test_result.rr
+    sds = arf.report_request
+    bench_source = sds.select_checklist!
+    [rs, OpenSCAP::Xccdf::Benchmark.new(bench_source).items]
```
As someone who's naive to what arf parsing is doing, it's not clear what rs, sds, or bench_source mean or what this method is returning.

isimluk replied:
Right. So perhaps, we can divide this into methods. If we come up with useful names for the methods, things will clear up.

There is a similar code for this in Sat 6.
https://github.com/OpenSCAP/scaptimony/blob/master/app/models/scaptimony/arf_report_raw.rb#L22

I have put blob handling into a separate model. Does it make sense to do something similar. i.e. have a mixin to BinaryBlob, that provides public `to_html` and can provide `benchmark_items` (yours meta) and `rule_results`?

moolitayer replied:
changed to the following, I hope that is better;
```
rule_results, benchmark_items = parse_arf(arf)

def parse_arf(arf)
  rule_results = arf.test_result.rr
  report_request = arf.report_request
  bench_source = report_request.select_checklist!
  [rule_results, OpenSCAP::Xccdf::Benchmark.new(bench_source).items]
end

```

jrafanie commented on app/models/openscap_result.rb:
```diff
@@ -0,0 +1,68 @@
+
+  def html
+    return unless binary_blob && binary_blob.binary
+    with_oscap_file(binary_blob.binary) do |arf_path|
```
Do we really need to write a temporary xml file to the filesystem to convert it to html?

isimluk replied:
We should be able to take ARF from memory. 

https://github.com/OpenSCAP/ruby-openscap/blob/master/lib/openscap/ds/arf.rb#L25
https://github.com/OpenSCAP/ruby-openscap/blob/master/lib/openscap/source.rb#L56

Something like this should just work:
```
OpenSCAP::DS::Arf.new(:content => raw_memory, :length => raw_memory_length, :path => 'my_arbitrary_arf_name.xml')
```

The `:path` you pass in, can be used by some error messages. However, a file at that location will not be touched. Some older version of OpenSCAP use the name to understand if you pass bzip stream (endswith(.xml.bz2)) or xml stream endswith('.xml'). The newer version of OpenSCAP do not care about filename, and just parse first few bytes to see, whether it is a bzip2.

moolitayer replied:
Done thanks!

jrafanie commented on app/models/openscap_result.rb:
```diff
@@ -0,0 +1,68 @@
+    self.binary_blob = BinaryBlob.new(:name => 'openscap_compliance_arf', :data_type => 'XML')
+    binary_blob.binary = value
+    with_oscap_file(value) do |arf_path|
+      create_results(arf_path)
```
Same as above, do we need to write a temp file here?  What is this block doing?

jrafanie commented on app/models/openscap_result.rb:
```diff
@@ -0,0 +1,68 @@
+  def create_results(arf_path)
+    results, meta = parse_arf(arf_path)
+
+    results.each do |k, v|
```
is `k` a `scap_rule`?  It would be helpful to readers like me to use a more descriptive set of variable names here.

moolitayer replied:
changed to |openscap_id, result|

jrafanie commented on app/models/openscap_result.rb:
```diff
@@ -0,0 +1,68 @@
+require 'openscap'
+require 'openscap/ds/arf'
+require 'openscap/xccdf/benchmark'
```
Does openscap have a loader that requires the files it needs?  It seems if openscap adds features, we'll need to keep this list of requires up to date.

moolitayer replied:
@jrafanie do you mean the binary it uses(oscap)? could this be working only because I have the binary installed?
If that is the case a dependency would have to be added on openscap-scanner. @isimluk is that correct?

jrafanie replied:
Sorry, @moolitayer, it's an observation that I didn't explain well.  What I mean is we have 3 requires for 66 lines of code.  If need to add more code here, will we need to add more requires?  If these requires are a common pattern, a loader in openscap (such as the require 'openscap' itself) could require them for us.  It doesn't have to happen, it's only an observation.

jrafanie commented on app/helpers/container_summary_helper.rb:
```diff
@@ -92,6 +92,33 @@ def textual_guest_applications
+  def textual_openscap_html
+    {
+      :link  => url_for(:controller => controller.controller_name, :action => 'openscap_html', :id => @record),
+      :value => _('1'),
```
What is 1 here?

moolitayer replied:
Changed the value of this label to 'HTML'

jrafanie commented on app/helpers/application_helper.rb:
```diff
@@ -202,6 +202,7 @@ def view_to_url(view, parent = nil)
     case view.db
     when "AdvancedSetting"  then association = "advanced_settings"
     when "CloudNetwork"     then association = "cloud_networks"
+    when "OpenscapRuleResult"       then association = "openscap_rule_results"
```
I know it's like this to begin with... :cry:

moolitayer replied:
@jrafanie didn't quite get that, do you want me to change something here

jrafanie replied:
Note, we don't have to fix this here.

It's just that this case statement is hard to read.  Most of these string conversions are `view.db.tableize`.
It's not clear to me why some are not following that convention but even then, these exceptions can be handled on their own.

moolitayer replied:
Addressed in #7259

jrafanie replied:
Thanks for looking into this.

simon3z commented on app/models/manageiq/providers/kubernetes/container_manager/scanning/job.rb:
```diff
@@ -124,12 +124,24 @@ def analyze
+    open_scap_result = OpenscapResult.new(:container_image => image)
+    open_scap_result.raw = image_inspector_client.fetch_oscap_arf
+
+    open_scap_result.save
```
@roliveri the rest of the information gathered by the smart state analysis are stored using `scan_metadata` and `sync_metadata`. Now, these OpenSCAP results seems not to respect the same behavior (directly saved from here), can you ACK, NACK this part? In case you're against it can you suggest the correct solution? Thanks.

roliveri replied:
@simon3z Having separate scan and sync phases would be more consistent with our existing SSA state machine, but given that the OpenSCAP scanning differs from standard SSA in a number of ways, I'm not sure if it matters enough to change it at this point. I say leave it the way it is, and we'll address any unforeseen issues if we encounter them.

simon3z commented:
This PR requires an explicit ACK/NACK from @roliveri about how to store the OpenSCAP data (`sync_metadata`-like?)

isimluk commented on app/models/openscap_result.rb:
```diff
@@ -0,0 +1,68 @@
+    return unless binary_blob && binary_blob.binary
+    with_oscap_file(binary_blob.binary) do |arf_path|
+      arf = OpenSCAP::DS::Arf.new(arf_path)
+      arf.html
```
please run `arf.destroy` here.

Ruby garbage collector will of course destroy the arf object. However, there is no way to tell GC that this object has special destructor.

moolitayer replied:
Done thanks!

moolitayer commented:
updated image: "Downloading the OpenSCAP HTML:"

moolitayer commented:
resource_id and resource_type columns removed from openscap_result creation script.

@simon3z @jrafanie please review

moolitayer commented:
two nits missing:
- show openscap content version(?)
- handle image show html with there  are no openscap results

@simon3z @jrafanie @h-kataria please review

moolitayer commented:
> two nits missing:
> 
> show openscap content version(?)

Done, see new screenshot above "This is how it looks if there are no openscap results"
The html text is now 'Available' or 'Not Available'
(@jrafanie had a comment on that earlier)

> handle image show html with there are no openscap results

Turns out for the kind of data we will have at this point (image CVEs) showing timestamp is the best we can do

Also changed variable name open_scap_result to correspond with class name openscap_result

@simon3z @jrafanie @h-kataria please review

simon3z commented on app/helpers/container_summary_helper.rb:
```diff
@@ -100,6 +100,35 @@ def textual_guest_applications
+  end
+
+  def textual_openscap_html
+    h = {}
```
Minor style suggestion (just taste).

```ruby
def textual_openscap_html
  h = {:label => _("OpenSCAP HTML")}
  if @record.openscap_result
    h[:value] = _('Available')
    h[:link] = url_for(:id => @record,
                       :controller => controller.controller_name,
                       :action => 'openscap_html')
  else
    h[:value] = _('Not Available')
  end
  h
end
```

simon3z commented on app/models/manageiq/providers/kubernetes/container_manager/scanning/job.rb:
```diff
@@ -124,12 +124,24 @@ def analyze
   end
 
+  def collect_compliance_data(image)
+    return unless image_inspector_client.respond_to?(:fetch_oscap_arf)
```
@moolitayer why this? if this is not implemented in the gem yet just comment the following code with a `TODO`

moolitayer replied:
Not sure I understand but if we merge without this check it will explode.

simon3z replied:
@moolitayer comment out the implementation of this method with a TODO. Or remove the entire call+method.

simon3z replied:
@moolitayer why is this? back? Just update the gem if it's available.

moolitayer replied:
@simon3z I need this code to run at tests when openscap is installed.

simon3z commented on spec/models/manageiq/providers/kubernetes/container_manager/scanning/job_spec.rb:
```diff
@@ -17,4 +67,125 @@
+
+  context "A single Container Scan Job," do
+    IMAGE_ID = '3629a651e6c11d7435937bdf41da11cf87863c03f2587fa788cf5cbfe8a11b9a'.freeze
+    IMAGE_NAME = 'test'.freeze
```
@moolitayer I see you added `freeze`, is it related to this PR?
(It's a good optimization but it will slip in unnoticed within this PR)

moolitayer replied:
> @moolitayer I see you added freeze, is it related to this PR?
> (It's a good optimization but it will slip in unnoticed within this PR)

removed freese, done in #7297 (I expect It will take some time to get to that so please ignore that for now)

simon3z commented on spec/models/manageiq/providers/kubernetes/container_manager/scanning/job_spec.rb:
```diff
@@ -17,4 +67,125 @@
+      allow_any_instance_of(described_class).to receive_messages(:kubernetes_client => MockKubeClient.new)
+      allow_any_instance_of(described_class).to receive_messages(
+        :image_inspector_client => MockImageInspectorClient.new(IMAGE_ID)
+      )
```
@moolitayer change of indentation?

simon3z commented on app/models/openscap_result.rb:
```diff
@@ -0,0 +1,66 @@
+require 'openscap/ds/arf'
+require 'openscap/xccdf/benchmark'
+
+class OpenscapResult < ApplicationRecord
```
@isimluk @shlomizadok can you review this? Thank you.

simon3z commented:
Small comments and waiting for @isimluk and @shlomizadok review. Overall LGTM :+1:

moolitayer commented:
ping @isimluk @shlomizadok

moolitayer commented:
ping @isimluk @shlomizadok

isimluk commented:
Sorry for being late @moolitayer. I think the code is fine.

Except we may want to do some extra wrapping because of memory leaks. Memory leaks aren't something you will see often in Ruby given the garbage collector. However, when we create bindings for two languages with different memory managements we just need to be extra careful. Ruby objects does not have a destructors, but things like Xccdf::Benchmark do.

I'll try to create pr against `moolitayer:persist_openscap` that will fix the memory leaks.

moolitayer commented:
> Sorry for being late @moolitayer. I think the code is fine.
> 
> Except we may want to do some extra wrapping because of memory leaks. Memory leaks aren't something you will see often in Ruby given the garbage collector. However, when we create bindings for two languages with different memory managements we just need to be extra careful. Ruby objects does not have a destructors, but things like Xccdf::Benchmark do.
> 
> I'll try to create pr against moolitayer:persist_openscap that will fix the memory leaks.

@isimluk Should ensure benchmark_items.[destroy](https://github.com/OpenSCAP/ruby-openscap/blob/master/lib/openscap/xccdf/benchmark.rb#L40) at [create_results](https://github.com/ManageIQ/manageiq/pull/6839/files#diff-a3e00569a5dbe8ceac40366d84487d89R39) satisfy this concern? thanks

isimluk commented:
To be 100% clear we need to 
```
benchmark.try(:destroy)
source_datastream.try(:destroy)
test_result.try(:destroy)
```
benchmark_items are owned (and disposed) by benchmark.

I have quickly crafted proposal at your repo https://github.com/moolitayer/manageiq/pull/1

isimluk commented:
ACK from OpenSCAP perspective.

jonnyfiveiq commented:
@moolitayer is it possible to access the SCAP detail from within Reporting? Eg I wish to be able to create a report just as screen shot "https://cloud.githubusercontent.com/assets/3010449/13555055/bf924c36-e3bf-11e5-8fdd-750dafadec3e.png" thanks

moolitayer commented:
> @moolitayer is it possible to access the SCAP detail from within Reporting? Eg I wish to be able to create a report just as screen shot "https://cloud.githubusercontent.com/assets/3010449/13555055/bf924c36-e3bf-11e5-8fdd-750dafadec3e.png" thanks

It is possible (on top of #7132). I am generating one now. Let me know if more info is needed

![screenshot from 2016-03-23 19-32-53](https://cloud.githubusercontent.com/assets/3010449/13994765/ff5a5d96-f12e-11e5-8eea-57fa4e6852ae.png)

moolitayer commented:
@simon3z please review

simon3z commented on app/models/manageiq/providers/kubernetes/container_manager/scanning/job.rb:
```diff
@@ -124,12 +124,24 @@ def analyze
+    # openscap_result = OpenscapResult.new(:container_image => image)
+    # openscap_result.raw = image_inspector_client.fetch_oscap_arf
+    #
+    # openscap_result.save
```
@roliveri I fear that when it will be the time what we'll implement this you won't ack it.

Can you please review it? Thanks cc @chessbyte

roliveri replied:
@simon3z I thought I commented on this a while ago, saying the current implementation was OK.

simon3z replied:
@roliveri sorry, it was probably hidden by the updates of the PR.

moolitayer commented:
rebase on top of #7259

simon3z commented:
Overall LGTM :+1:

@roliveri and @isimluk already ack'ed this. So for as much as I am concerned we can merge.

cc @blomquisg @chessbyte it all depends on how much you're strict on the fact that `collect_compliance_data` is currently not implemented as we're waiting for the gem to be ready.

moolitayer commented:
Travis failure: it seems that requiring the gem isn't enough, we need the binary too?

```
Failure/Error: require 'openscap'
     
     LoadError:
       Could not open library 'libopenscap.so.8': libopenscap.so.8: cannot open shared object file: No such file or directory.
       Could not open library 'openscap': openscap: cannot open shared object file: No such file or directory.
       Could not open library 'libopenscap.so': libopenscap.so: cannot open shared object file: No such file or directory
```

simon3z commented:
> Travis failure: it seems that requiring the gem isn't enough, we need the binary too?

Yes, that's what @isimluk mentioned since the beginning. The rendering is done by the binary.

Fryguy commented:
That's going to be a hassle on travis and probably mac dev machines.

moolitayer commented:
> That's going to be a hassle on travis and probably mac dev machines.

I'm open to any options of an optional dependency (tough travis should probably always test it).

@Fryguy @jrafanie could you point me in the right direction for adding this dependency?
I need to get it ASAP. A checklist of other related items I need to do is also appreciated. (integration, other people that need to be aware of such change, other stream builds etc.)

moolitayer commented:
rebased on #7502

jrafanie commented:
> @Fryguy @jrafanie could you point me in the right direction for adding this dependency?
> I need to get it ASAP. A checklist of other related items I need to do is also appreciated. (integration, other people that need to be aware of such change, other stream builds etc.)

@moolitayer having it installed on travis is probably not a viable option in the short term.  @Fryguy has laid it out on gitter.  Let us know if you need help.

moolitayer commented:
Since it is not viable to have the binary (since Travis is Ubuntu) these are the steps I took:

- removed the tests I wrote that test the openscap code :cry: as well as the example arf file. They test the wrapper scap code that rely heavily on the binary. It is not viable to mock the whole thing (unless I was trying to test the wrapper itself) 
- wrapped the require code that raises exceptions in unrelated code calling openscap_result when the binary isn't available (container_image_controller_spec.rb) the require will now only be called when there are actual result which will never be the case at Travis.
-  at the scanning/job.rb, return the code ```return unless image_inspector_client.respond_to?(:fetch_oscap_arf)``` @simon3z I know you didn't like it before, but it now allows to avoid running until feature is merged and also it will never be executed on Travis since it is not mocked. Let me know if you still don't like it.
- The refactor to job_spec.rb is now longer relevant to this pr and has been resubmitted in #7546

@simon3z @jrafanie @chessbyte please review/merge,
thanks

simon3z commented:
@moolitayer @isimluk @Fryguy @jrafanie since the scope here is not testing the OpenSCAP gem functionalities (@isimluk I assume you're testing it somewhere else) , I think that it can be fully mocked (just a couple of methods) for travis tests.

This obviously won't cover the case where the OpenSCAP gem will change its API (because we're mocking the behavior).
(After all this is what happens with VCR, etc.)

Maybe we can have additional optional tests that run only if the OpenSCAP binary is available.

simon3z commented on app/models/openscap_result.rb:
```diff
@@ -0,0 +1,77 @@
+    string.to_s.encode('utf-8', :invalid => :replace, :undef => :replace, :replace => '_')
+  end
+
+  def with_openscap_arf(raw)
```
@isimluk if you have something similar in Foreman you may consider to move this inside your ruby lib.

isimluk replied:
Sure, feel free to file an issue against rubygem-openscap.

So far rubygem-openscap is just lean wrapper around libopenscap.so. The value of having lean wrapper is that it is clear what kind of API it should provide. I am afraid that when we put these few lines helper into the gem, we risk for things to be less flexible (so far all the rubygem-openscap releases are backward API compatible).

jrafanie commented:
> Maybe we can have additional optional tests that run only if the OpenSCAP binary is available.

Agreed, I believe @carbonin and @Fryguy provided ideas on how we are doing similar behavior for testing of pglogical/replication.  If you setup your environment to have the correct gems/libraries, you should be able to run the tests with minimal changes.  It would be nice if the tests detected if those tests are supported... But yes, having just a basic integration test would be sufficient since we don't want to duplicate testing that is happening on the OpenSCAP gem itself.

carbonin commented:
LinuxAdmin should be able to check for the presence of a particular rpm ([`LinuxAdmin::Rpm.list_installed`](https://github.com/ManageIQ/linux_admin/blob/master/lib/linux_admin/rpm.rb#L9-L15)) and as @jrafanie said we did a similar thing with the replication tests [here](https://github.com/ManageIQ/manageiq/blob/master/spec/replication/replication_spec.rb#L28)

moolitayer commented:
@simon3z @jrafanie @carbonin thanks for the comments

- job_spec.rb now tests code & gem/binary usage if binary is installed
- new openscap_result_spec.rb tests basic flow regardless of binary

Please review

ps @carbonin I preferred to test for the binary as you've done do at [linux_admin](https://github.com/ManageIQ/linux_admin/blob/master/lib/linux_admin/scap.rb#L8)
to test for the binary directly and not the package.

moolitayer commented:
@dclarizio @h-kataria is this ready UI-wise?

h-kataria commented on app/models/manageiq/providers/kubernetes/container_manager/scanning/job.rb:
```diff
@@ -124,12 +124,22 @@ def analyze
   end
 
+  def collect_compliance_data(image)
+    # openscap_result = OpenscapResult.new(:container_image => image)
```
@moolitayer is this intentional to have all lines in this method commented?

moolitayer replied:
Yes. The intention is to remove the comments when a new image_inspector_client is available. See also [this](https://github.com/ManageIQ/manageiq/pull/6839#issuecomment-200556024) comment

h-kataria replied:
@mmolitayer i missed the other comment in pool of comments. Looks good.

h-kataria commented:
@dclarizio UI changes look good.

blomquisg commented on app/models/openscap_result.rb:
```diff
@@ -0,0 +1,77 @@
+
+  def html
+    raw = binary_blob.try(:binary)
+    return unless raw
```
Is it safe for this method to return `nil`?

It looks like this is [sent to `send_data` above](https://github.com/ManageIQ/manageiq/pull/6839/files#diff-12a48123abd31efa3db8f18294aeec46R20).  Wondering if that's fine, or if the caller or this method should check for `nil` and send back an appropriate response.

moolitayer replied:
I Assume that if the (image has an openscap_result)[https://github.com/ManageIQ/manageiq/pull/6839/files#diff-78122ac21511d34c71d004ea16f3faacR123] it must have a valid html. I've changed the code to reflect that.

blomquisg commented on app/models/openscap_result.rb:
```diff
@@ -0,0 +1,77 @@
+    self.binary_blob = BinaryBlob.new(:name => 'openscap_compliance_arf', :data_type => 'XML')
+    binary_blob.binary = value
+    with_openscap_objects(binary_blob.binary) do |rule_results, benchmark_items|
+      create_results(rule_results, benchmark_items)
```
Hrm, should this actually create a whole new set of results each time `raw=` is called?

For instance, if I have the following calls:

```ruby

result = OpenscapResult.first
result.raw = a_binary_blob       # two rule results in this blob
result.raw = another_binary_blob # three rule results in this blob
```

It seems like I'm gonna have the following:

```ruby
result.raw == another_binary_blob

# two from the first assignment and three from the second
result.openscap_rule_results.size == 5 
```

Maybe I read something wrong though?

moolitayer replied:
That is right @matthewd 
Per image I always:
- Destroy the old openscap_result (and openscap_rule_results due to :autosave => true)
- Create a new one
- Set raw once

Looks like the example above would leave dangling openscap_rule_results. I added a check at raw= to make sure that does not happen
```
raise "Cannot invoke raw setter more than once." unless binary_blob.nil?
```

simon3z commented:
There are some questions from @blomquisg to address (reply or fix).

Other than that LGTM :+1: ready for merge cc @chessbyte @blomquisg

moolitayer commented:
- Addressed comments comments from @blomquisg 
- Added ``` image.openscap_result.try(:destroy) ``` to remove old openscap_results 
- Removed blank line

moolitayer commented:
Needed to fix tests as well.

durandom commented on app/models/manageiq/providers/kubernetes/container_manager/scanning/job.rb:
```diff
@@ -124,12 +124,23 @@ def analyze
   end
 
+  def collect_compliance_data(image)
+    # image.openscap_result.try(:destroy)
```
I see that you want to un-comment this code once 
> a new image_inspector_client is available. See also this comment

which is fine.
But how about moving this bit to a subsequent PR? Maybe the gem never gets ready or the api changes?

At least I would expect some explanation inside the code why this is commented out.

moolitayer replied:
@durandom I've added a comment. 
The commented code Is nice for me since I can run tests at my local machine by only un commenting the code[1], but I guess I can remove it at this point @simon3z ?

[1] As Travis does not, and will never have the needed binary for test on that specific code.

durandom commented on app/models/container_image.rb:
```diff
@@ -13,6 +13,8 @@ class ContainerImage < ApplicationRecord
   has_many :guest_applications, :dependent => :destroy
   has_one :computer_system, :as => :managed_entity, :dependent => :destroy
   has_one :operating_system, :through => :computer_system
+  has_one :openscap_result
```
shouldnt this be `:dependent => :destroy` ?

moolitayer replied:
> shouldnt this be :dependent => :destroy ?

Right, Thanks, Fixed

simon3z commented:
@moolitayer this needs rebase

durandom commented on db/migrate/20160220123313_add_open_scap_result_entities.rb:
```diff
@@ -0,0 +1,23 @@
+class AddOpenScapResultEntities < ActiveRecord::Migration
+  def up
```
This can be wrapped in a `def change` migration. E.g. https://github.com/ManageIQ/manageiq/blob/c0868248d2681c6cd6644fc8c05e11b4f255cf4f/db/migrate/20160211113430_create_cloud_databases.rb

moolitayer replied:
@durandom fixed. Do the indexes look fine?

durandom commented on db/migrate/20160220123313_add_open_scap_result_entities.rb:
```diff
@@ -0,0 +1,23 @@
+class AddOpenScapResultEntities < ActiveRecord::Migration
+  def up
+    create_table :openscap_results do |t|
+      t.belongs_to :container_image, :type => :bigint
```
this also automatically creates an index if you use `t.belongs_to` or `t.references`

durandom commented on app/models/manageiq/providers/kubernetes/container_manager/scanning/job.rb:
```diff
@@ -124,12 +124,24 @@ def analyze
+    # TODO: Un comment this code once an image_inspector_client that knows to fetch_oscap_arf arrives.
+    # image.openscap_result.try(:destroy)
+    # openscap_result = OpenscapResult.new(:container_image => image)
+    # openscap_result.raw = image_inspector_client.fetch_oscap_arf
```
How about guarding the method with
```ruby
def collect_compliance_data(image)
  return unless image_inspector_client.respond_to? :fetch_oscap_arf
  ...
end

jrafanie replied:
:+1: I agree, if possible, let existing non-conforming clients quietly return.  When the clients know how to do it, then, and only then, do it.

moolitayer replied:
I had it before @simon3z does not like it.

@simon3z I'm pushing it, let me know if you object.

moolitayer replied:
The order in which I delete and add is not right since I will lose old if there is a problem building new. Is this better? @jrafanie @durandom 
```
def collect_compliance_data(image)
    return unless image_inspector_client.respond_to? :fetch_oscap_arf
    openscap_result = OpenscapResult.new(:container_image => image)
    openscap_result.raw = image_inspector_client.fetch_oscap_arf
    image.openscap_result.try(:destroy) # delete old
    openscap_result.save
end
```

durandom commented on app/models/openscap_result.rb:
```diff
@@ -0,0 +1,76 @@
+  end
+
+  def raw=(value)
+    raise "Cannot invoke raw more than once." unless binary_blob.nil?
```
why can `raw=` be called only once?

I think you can express what you actually want better with validations
```ruby
validates :binary_blob, presence: true
```

Or do you really want to make sure, that once a blob and the resulting rule_results are saved they cannot be changed anymore?

blomquisg replied:
I think what you want here is a couple of things:

1. @durandom's suggestion about `validates :binary_blob, presence: true`
2. possibly move the creation of the child `openscap_rule_results` to a `before_save`
3. make `raw` readonly so it can only be set when the `openscap_result` object is new

That way, it doesn't matter how many times `raw` is called on a new `openscap_result` object, it only gets exploded once during save, and once it's save it's immutable.

Does that sound about right?  Or, am I missing the point?

durandom replied:
1. can also be added to the association `has_one :binary_blob, required: true` 
2. makes sense, if `OpenscapResult` is created only once and probably never updated. Otherwise I would add some logic to check if the blob has changed. But having it always create the rules once the blob changes makes sense
3. I think 1&2 remove the need for `raw=` 

ad 3. you might want to have a similar method, named `attach_raw_result(openscap_xml)` which hides the BinaryBlob

durandom commented on app/models/openscap_result.rb:
```diff
@@ -0,0 +1,76 @@
+    end
+  end
+
+  def html # Must be called after setting raw or it will explode
```
I'd remove the comment and add a guard clause
```
return '' unless binary_blob.present?
```

moolitayer replied:
Is it wrong to want it to explode if there is no binary_blob?
I do not want to handle a returned '' and it should never happen (after adding :required => true as you suggested I think It can only happen in an object not yet saved)

durandom commented on app/models/openscap_result.rb:
```diff
@@ -0,0 +1,76 @@
+  end
+
+  def with_openscap_arf(raw)
+    raise "no block given" unless block_given?
```
not needed, ruby will complain and raise `LocalJumpError` if you call the method without a block

durandom commented on app/models/openscap_result.rb:
```diff
@@ -0,0 +1,76 @@
+  has_one    :binary_blob,           :dependent => :destroy, :autosave => true, :as => :resource
+  has_many   :openscap_rule_results, :dependent => :destroy, :autosave => true
+
+  def self.openscap_available?
```
How about moving the openscap requirement into its own class?
By default you just delegate the methods and in a testing env you give it the stub.
This way you could also split the tests into a) testing OpenscapResult and b) testing the correct use of the openscap gem
Especially b) is a problem with mocks, as @simon3z pointed out

I just recently came a long a good blog post about this:
http://blog.plataformatec.com.br/2016/02/experimenting-with-explicit-contracts-with-ruby/

moolitayer replied:
> How about moving the openscap requirement into its own class?
> By default you just delegate the methods and in a testing env you give it the stub.
> This way you could also split the tests into a) testing OpenscapResult and b) testing the correct use of the openscap gem
> Especially b) is a problem with mocks, as @simon3z pointed out
> 
> I just recently came a long a good blog post about this:
> http://blog.plataformatec.com.br/2016/02/experimenting-with-explicit-contracts-with-ruby/

@durandom I'm not sure I fully understand. Will the above allow for both environments that have the binary (my computer) and those that do not (Travis)?
I need both (which I currently support)

moolitayer commented:
- Addressed comments

@durandom @blomquisg  @jrafanie Thanks for the helpful comments

blomquisg commented:
> Addressed comments
> @durandom @blomquisg @jrafanie Thanks for the helpful comments!

I like this a lot better.  Thanks for trudging through it with us!

blomquisg commented:
And, looks like we have [UI sign off, too](https://github.com/ManageIQ/manageiq/pull/6839#issuecomment-205404281).

Waiting for green...

moolitayer commented:
> And, looks like we have UI sign off, too.
> 
> Waiting for green...

**[breathing heavily]**

jrafanie commented:
LGTM

durandom commented:
@moolitayer good job! :clap:

moolitayer commented:
@durandom Thanks!
